File: src/main/java/systems/conduit/main/mixins/item/ItemStackMixin.java
package systems.conduit.main.mixins.item;

import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.Pose;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.context.UseOnContext;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
import systems.conduit.main.Conduit;
import systems.conduit.main.api.mixins.ServerPlayer;
import systems.conduit.main.core.events.types.PlayerEvents;
import systems.conduit.main.core.events.types.WorldEvents;

import java.util.function.Consumer;

@Mixin(value = ItemStack.class, remap = false)
public class ItemStackMixin {

    @Inject(method = "useOn", at = @At("HEAD"))
    private void useOn(UseOnContext context, CallbackInfoReturnable<InteractionResult> cir) {
        boolean isCrouching = (context.getPlayer() != null && context.getPlayer().getPose() != null && context.getPlayer().getPose() == Pose.values()[5]);
        WorldEvents.BlockInteractEvent event = new WorldEvents.BlockInteractEvent((ServerPlayer) context.getPlayer(),
                context.getLevel().getBlockState(context.getClickedPos()), context.getHand(), context.getItemInHand(),
                context.getClickedFace(), isCrouching, context.isInside());
        Conduit.getEventManager().dispatchEvent(event);
    }

    @Inject(method = "hurtAndBreak", at = @At(value = "HEAD", target = "Lnet/minecraft/world/item/ItemStack;hurt(ILjava/util/Random;Lnet/minecraft/server/level/ServerPlayer;)Z"), cancellable = true)
    public void hurtAndBreak(int i, LivingEntity t, Consumer<? extends LivingEntity> consumer, CallbackInfo ci) {
        if (t instanceof ServerPlayer) {
            PlayerEvents.ItemBreakEvent event = new PlayerEvents.ItemBreakEvent((ServerPlayer) t, (ItemStack) ((Object) this));
            Conduit.getEventManager().dispatchEvent(event);

            if (event.isCanceled()) ci.cancel();
        }
    }
}

File: src/main/kotlin/systems/conduit/core/mixins/event/ItemStackMixin.kt
package systems.conduit.core.mixins.event

import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.Pose
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.UseOnContext
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import systems.conduit.api.ServerPlayer
import systems.conduit.core.Conduit
import systems.conduit.core.events.types.WorldEvents

@Mixin(value = [ItemStack::class], remap = false)
class ItemStackMixin {

    @Inject(method = ["useOn"], at = [At("HEAD")])
    private fun useOn(context: UseOnContext, cir: CallbackInfoReturnable<InteractionResult>) {
        val isCrouching = context.player != null && context.player!!.pose != null && context.player!!.pose == Pose.values()[5]
        val event = WorldEvents.BlockInteractEvent(context.player as ServerPlayer,
                context.level.getBlockState(context.clickedPos), context.hand, context.itemInHand,
                context.clickedFace, isCrouching, context.isInside)
        Conduit.eventManager.dispatchEvent(event)
    }
}
